src/super_gradients/training/utils/pose_estimation/dekr_decode_callbacks.py

```diff
@@ -0,0 +1,303 @@
+from typing import Union, Tuple, List
+
+import numpy as np
+import torch
+from torch import nn, Tensor
+
+
+def get_locations(output_h: int, output_w: int, device):
+    """
+    Generate location map (each pixel contains its own XY coordinate)
+
+    :param output_h: Feature map height (rows)
+    :param output_w: Feature map width (cols)
+    :param device: Target device to put tensor on
+    :return: [H * W, 2]
+    """
+    shifts_x = torch.arange(0, output_w, step=1, dtype=torch.float32, device=device)
+    shifts_y = torch.arange(0, output_h, step=1, dtype=torch.float32, device=device)
+    shift_y, shift_x = torch.meshgrid(shifts_y, shifts_x, indexing="ij")
+    shift_x = shift_x.reshape(-1)
+    shift_y = shift_y.reshape(-1)
+    locations = torch.stack((shift_x, shift_y), dim=1)
+
+    return locations
+
+
+def get_reg_poses(offset: Tensor, num_joints: int):
+    """
+    Decode offset predictions into absolute locations.
+
+    :param offset: Tensor of [num_joints*2,H,W] shape with offset predictions for each joint
+    :param num_joints: Number of joints
+    :return: [H * W, num_joints, 2]
+    """
+    _, h, w = offset.shape
+    offset = offset.permute(1, 2, 0).reshape(h * w, num_joints, 2)
+    locations = get_locations(h, w, offset.device)
+    locations = locations[:, None, :].expand(-1, num_joints, -1)
+    poses = locations - offset
+
+    return poses
+
+
+def offset_to_pose(offset, flip=False, flip_index=None):
+    """
+    Decode offset predictions into absolute locations.
+
+    :param offset: [1, 2 * num_joints, H, W]
+    :param flip: True to decode offsets for flipped keypoints (WHen horisontal flip TTA is used)
+    :param flip_index: Indexes of keypoints that must be swapped to account for horizontal flip
+    :return: [1, 2 * num_joints, H, W]
+    """
+    num_offset, h, w = offset.shape[1:]
+    num_joints = int(num_offset / 2)
+    reg_poses = get_reg_poses(offset[0], num_joints)
+
+    if flip:
+        reg_poses = reg_poses[:, flip_index, :]
+        reg_poses[:, :, 0] = w - reg_poses[:, :, 0] - 1
+
+    reg_poses = reg_poses.contiguous().view(h * w, 2 * num_joints).permute(1, 0)
+    reg_poses = reg_poses.contiguous().view(1, -1, h, w).contiguous()
+
+    return reg_poses
+
+
+def hierarchical_pool(heatmap, pool_threshold1=300, pool_threshold2=200):
```

Comment:
There are a few methods here missing docs...
Please add/make them "private" if they're for internal use.